### PR TITLE
feat(event-stream): quiesce + drain shutdown ordering (BD-2a)

### PR DESCRIPTION
**Pull request by H31MDALLR**
## Summary

First implementation PR off the BD-2 design RFC (#795). Adds the two shutdown-ordering primitives the headless-exit race requires:

- **`quiesce!`** fences future publishers for a workflow id. After return, `publish!` for that workflow returns `{:rejected? true :reason :workflow-quiesced ...}` instead of running sinks or subscribers. The terminal `:workflow/completed` / `:workflow/failed` event becomes the genuine last publish for that workflow.
- **`drain!`** waits for in-flight publishes to settle, then asks each sink that exposes a `:drain` hook under metadata to flush. Synchronous sinks (file, stdout, stderr) need no hook; the fleet sink's batched HTTP path is the motivating exception. Returns a structured result so callers can map to a non-zero exit.

## Wire-up

`bases/cli/workflow_runner.clj` `run-workflow!`:

- After `execute-workflow-pipeline` returns and `close-artifact-store` has run, call `quiesce!` for the workflow then `drain!`. On non-OK drain in normal mode, throw an `ex-info` carrying the drain result so the existing catch path renders the failure and the CLI exits non-zero.
- `MINIFORGE_BEST_EFFORT_SHUTDOWN=1` is the documented escape hatch for local/dev loops: drain failures still log a structured warning to stderr but the run continues. Per the BD-2 RFC v4 commitment, this mode is explicit and noisy — never silent.
- Drain result is attached to the run result as `:event-durability` so downstream callers can detect a degraded shutdown.

## Internal change to `publish!`

- Snapshot `:quiesced-workflows` once on entry; reject before any side effect, so a quiesced workflow's publish never runs sinks or subscribers.
- Increment `:in-flight` after the quiesce check, decrement in a finally so a sink exception still releases the slot.

## Tests

`components/event-stream/test/.../quiesce_drain_test.clj` pins the contract:

- quiesce rejects future publishes for the named workflow;
- quiesce only fences the named workflow (independent workflows continue);
- quiesce without a workflow id is a global in-flight barrier only;
- drain returns ok immediately when all sinks are synchronous;
- drain invokes the `:drain` metadata hook when present;
- drain reports structured `:sink-error` when a sink fails or throws;
- the explicit "publish-during-drain does not bypass quiesce" test pins the v1 race the primitives are meant to close;
- in-flight tracking decrements on sink exception so drain can't hang on a stuck counter.

## Out of scope (tracked for follow-on)

- **File-sink atomic write + fsync hardening** (BD-2a test #8 in the RFC). Current file sink uses `spit`, which closes the writer but does not fsync. Bigger contained change; lands separately so this PR stays focused on the publisher-fence + sink-drain contract surface.
- **CLI exit-code wiring beyond `throw`**. The current PR throws on non-OK drain; whatever calls `run-workflow!` exits non-zero on the rethrown exception. A more structured `event_durability` field in the run result is in place; richer CLI exit-code handling is a small follow-on.

## Validation

```
bb pre-commit
```

All checks passed. 5208 tests, 0 failures (the 1 flaky failure observed during the BD-2 RFC commit was unrelated and didn't reproduce here).

## Test plan

- [ ] reviewer agrees the metadata-based drain hook is the right opt-in for sinks that need flushing (vs an explicit protocol).
- [ ] reviewer agrees throwing-on-non-OK-drain is the right wiring vs returning a structured result and trusting the caller.
- [ ] reviewer confirms the file-sink fsync hardening is appropriate to defer.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: components/event-stream/src/ai/miniforge/event_stream/core.clj [hunk 1]
```diff
@@ -160,6 +190,113 @@
...
+            :or   {timeout-ms 5000}}]
+   (when workflow-id
+     (swap! stream update :quiesced-workflows conj workflow-id))
+   (let [deadline (+ (System/currentTimeMillis) timeout-ms)
+         settled? (wait-for-condition #(zero? (:in-flight @stream)) deadline)
+         pending  (:in-flight @stream)]
+     (if settled?
+       {:ok? true :pending-publishers 0}
+       {:ok? false :reason :timeout :pending-publishers pending}))))
+
```
[the remark this answers is not in the document]
Comment (H31MDALLR) [reply components/event-stream/src/ai/miniforge/event_stream/core.clj:322]: Fixed in a5c51cd6. quiesce! now reads in-flight once at return time and reports the same value in both ok and timeout branches. The contract is observational: ok=true means in-flight reached zero at some point during the wait; pending is the value at return.

File: components/event-stream/src/ai/miniforge/event_stream/core.clj [hunk 2]
```diff
@@ -160,6 +190,113 @@
...
+             results (mapv
+                      (fn [sink]
+                        (if-let [drain-fn (:drain (meta sink))]
+                          (let [remaining (max 100 (- deadline (System/currentTimeMillis)))]
+                            (try
+                              {:sink sink :result (drain-fn {:timeout-ms remaining})}
+                              (catch Exception e
+                                {:sink sink :result {:ok? false
+                                                     :reason :sink-error
+                                                     :error (ex-message e)}})))
```
[the remark this answers is not in the document]
Comment (H31MDALLR) [reply components/event-stream/src/ai/miniforge/event_stream/core.clj]: Fixed in a5c51cd6. drain! switched from mapv with a (max 100 …) floor to a reduce that checks (- deadline now) <= 0 before each sink and short-circuits remaining sinks with :reason :timeout. New regression test drain-budget-short-circuits-remaining-sinks-when-deadline-passes asserts a slow first sink does not grant later sinks any extra wall time.

File: components/event-stream/src/ai/miniforge/event_stream/core.clj [hunk 3]
```diff
@@ -99,47 +99,77 @@
...
-                                     :event-type (:event/type event)
-                                     :anomaly anomaly}}))))))))
-    (when logger
-      (log/debug logger :event-stream :event/published
-                 {:message "Event published"
-                  :data {:event-type (:event/type event)
-                         :workflow-id workflow-id
-                         :sequence (:event/sequence-number event)}}))
-    event))
+  (let [workflow-id (:workflow/id event)
```
Comment (H31MDALLR) [inline components/event-stream/src/ai/miniforge/event_stream/core.clj]: This is far too complex of a function. Refactor, pull out functions until it looks simple and composes like a pipeline, and then make sure those functions are at the right strata and tested.

Comment (H31MDALLR) [reply components/event-stream/src/ai/miniforge/event_stream/core.clj]: Refactored in af908c1b. `publish!` is now:

```clojure
(defn publish! [stream event]
  (or (rejection-if-quiesced stream event)
      (with-in-flight stream
        (fn []
          (let [{:keys [sinks subscribers filters logger]} @stream]
            (deliver-to-sinks! sinks event logger)
            (record-event! stream event)
            (deliver-to-subscribers! subscribers filters event logger)
            (log-published! logger event)
            event)))))
```

The body is the pipeline you asked for. Underlying helpers, all Layer 0 `defn-`:

- `workflow-quiesced?` — pure check against the stream's quiesced set.
- `rejection-result` / `log-rejection!` / `rejection-if-quiesced` — quiesce-rejection branch, composed.
- `with-in-flight` — increment / try / finally / decrement. The cross-cutting concern that lets `quiesce!` / `drain!` observe at-rest.
- `record-event!` — append to in-memory log.
- `deliver-to-sink!` / `deliver-to-sinks!` — single sink with isolation, then fanout.
- `deliver-to-subscriber!` / `deliver-to-subscribers!` — single subscriber with filter + isolation, then fanout.
- `log-published!` — debug log on successful publish.

`publish!` itself is at Layer 1 alongside the rest of the bus operations; helpers sit at Layer 0 (lower stratum, no upward dependencies).

External behavior is unchanged. Existing integration tests in `core_test.clj` and `quiesce_drain_test.clj` still exercise the composed path. New `publish_helpers_test.clj` adds per-stratum coverage: 13 tests covering boundary conditions for each helper, including the in-flight slot release on body exception, sink/subscriber failure isolation, filter respect, and the rejection-result shape.
